Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(S,R)`-bimodule.
+Consider the tensor functor `(X ⊗[R] ·)` from the category of left `R`-module to the category of
+left `S`-module and the hom functor `(X →ₗ[S] ·)` from the category of left `S`-module to the
+category of left `R`-module. They form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_S(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
```

Comment:
No, this is definitely not more natural, and in fact doesn't generalize as well!

The right way to think about order when `Hom` is involved is as $\mathsf{Hom}(X, Y) \cong X^* \otimes Y$. Thus the "natural" thing to say is:

$$\mathsf{Hom}(X \otimes Y \to Z) \cong (X \otimes Y)^* \otimes Z \cong Y^* \otimes X^* \otimes Y \cong \mathsf{Hom}(Y \to \mathsf{Hom}(X \to Z)).$$

This is the thing that keeps working in the most noncommutative settings.

Reply:
This makes sense! Thanks